src/box/lua/config/applier/credentials.lua

```diff
@@ -819,49 +896,21 @@ local wait_sync
 -- {type = obj_type, name = obj_name} - perform a per-object sync, skip
 --                                      return message
 local function sync_credentials_worker()
+    fiber.self():name('sync_credentials', { truncate = true })
+
     while true do
         local obj_to_sync = sync_tasks:get()
 
-        if obj_to_sync.type == 'BLOCKING_FULL_SYNC' or
-                obj_to_sync.type == 'BACKGROUND_FULL_SYNC' then
-
-            if box.info.ro then
-                log.verbose('credentials: the database is in the read-only ' ..
-                    'mode. Waiting for the read-write mode to set up the ' ..
-                    'credentials given in the configuration. If another ' ..
-                    'instance in the replicaset will write these new ' ..
-                    'credentials, this instance will receive them and will ' ..
-                    'skip further actions.')
-                box.ctl.wait_rw()
-            end
-
-            -- On config reload, drop the old list of registered objects.
-            target_object_map = {
-                ['role'] = {},
-                ['space'] = {},
-                ['function'] = {},
-                ['sequence'] = {},
-            }
+        local _, err = pcall(sync_object, obj_to_sync)
 
-            local credentials = get_credentials(config)
-
-            register_objects(credentials.roles)
-            register_objects(credentials.users)
-
-            box.atomic(function()
-                create_roles(credentials.roles)
-                create_users(credentials.users)
-
-                sync_privileges(credentials)
-            end)
-
-            if obj_to_sync.type == 'BLOCKING_FULL_SYNC' then
+        if obj_to_sync.type == 'BLOCKING_FULL_SYNC' then
+            if err then
+                wait_sync:put(err)
+            else
                 wait_sync:put('Done')
             end
-        else
-            local credentials = get_credentials(config)
-
-            box.atomic(sync_privileges, credentials, obj_to_sync)
+        elseif err then
+            error(err)
```

Comment:
Should we really terminate the fiber in this case? Maybe it is better to log an error and pick up the next task?

Reply:
You are right. That's not the best way to deal with the worker thread. Fixed.